### PR TITLE
Library Config File

### LINKED ISSUE #1084: [TASK] Plain-text Library Dependencies
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
In preparation for #1055, RAVEN needs a simplifed tracking system for its dependent libraries.

**Describe the solution you'd like**
A plaintext or config-style file instead of `RavenUtils.py` is in order, along with a simplified parser.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [ ] 1. If the issue is a defect, is the defect fixed?
- [ ] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [ ] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [ ] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [ ] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #1084 

##### What are the significant changes in functionality due to this change request?
Removes the "RavenUtils" module, replacing it with two files:
- `raven/dependencies.ini`, a config file with Python dependencies categorized by section, and
- `raven/scripts/library_handler.py`, which compiles the library sets to use and returns completed conda/pip commands.

The old command line arguments to RavenUtils were quite hard to follow, so those have been retooled a bit for maintainable clarity; no functionality has been changed.

That said, we now always report being in python3 since we're not python2 compatible.

The new structure should facilitate a plugin-based library extension much more easily than the old system.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [ ] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: dependencies.ini [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+[mac]
+# nothing currently
+
+[linux]
+# these prevent Intel crash errors with mkl
+nomkl
+numexpr
+
+##### Alternate installs
+# install using pip
```
Comment (wangcj05) [inline dependencies.ini]: Could you add more descriptions about how we formulate the list of libraries for different situations? For example, when use Conda to install: we need 'core', 'forge', and 'pip'. For pip-install, we need 'core', 'forge', and 'pip', but remove 'pip-install'

**Comment by wangcj05** [comment]
@PaulTalbot-INL I think this new structure very much.

File: dependencies.ini (hunk 1 above)
Comment (PaulTalbot-INL) [reply dependencies.ini]: I tried to capture that with the `##### Typical Installation` and `##### Alternate installs` structure, then `core, secondary, tertiary` and `optional, windows, mac, linux`. Let me know how you think this could be clarified.

I feel any additional logic should be contained in the `library_handler.py`; otherwise, it has to be maintained in two places. Let me know if more comments should be added to clarify that.

Comment (PaulTalbot-INL) [reply dependencies.ini]: I went ahead and added some comments anyway; let me know if that's along the lines you were thinking.

**Comment by PaulTalbot-INL** [comment]
Tests are passing, could someone review?

File: tests/framework/CodeInterfaceTests/Neutrino/Neutrino_Base/sampleNeutrino/1/NeutrinoInput.nescene [hunk 2]
```diff
@@ -27,9 +27,9 @@
...
             <property name="ServerPort" hidden="0" locked="0" val="20200"/>
-            <property name="SceneFilePath" hidden="0" locked="0" val="C:\msys64\home\RYANED\Emerald\projects\raven\tests\framework\CodeInterfaceTests\Neutrino\Neutrino_Base\sampleNeutrino\1\NeutrinoInput.nescene"/>
-            <property name="SaveDir" hidden="0" locked="0" val="C:\msys64\home\RYANED\Emerald\projects\raven\tests\framework\CodeInterfaceTests\Neutrino\Neutrino_Base\sampleNeutrino\1\"/>
-            <property name="CacheDir" hidden="0" locked="0" val="C:\msys64\home\RYANED\Emerald\projects\raven\tests\framework\CodeInterfaceTests\Neutrino\Neutrino_Base\sampleNeutrino\1\"/>
+            <property name="SceneFilePath" hidden="0" locked="0" val="/Users/talbpw/raven/tests/framework/CodeInterfaceTests/Neutrino/Neutrino_Base/sampleNeutrino/1/NeutrinoInput.nescene"/>
+            <property name="SaveDir" hidden="0" locked="0" val="/Users/talbpw/raven/tests/framework/CodeInterfaceTests/Neutrino/Neutrino_Base/sampleNeutrino/1/"/>
+            <property name="CacheDir" hidden="0" locked="0" val="/Users/talbpw/raven/tests/framework/CodeInterfaceTests/Neutrino/Neutrino_Base/sampleNeutrino/1/"/>
```
Comment (wangcj05) [inline tests/framework/CodeInterfaceTests/Neutrino/Neutrino_Base/sampleNeutrino/1/NeutrinoInput.nescene]: why the unique absolute path is used here? very confusing for me.

Comment (wangcj05) [reply tests/framework/CodeInterfaceTests/Neutrino/Neutrino_Base/sampleNeutrino/1/NeutrinoInput.nescene]: Same for all NeurinoInput

File: scripts/library_handler.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+  for lib, needVersion in need.items():
+    # some libs aren't checked from within python
+    if lib in skipChecks:
+      continue
+    found, msg, foundVersion = checkSingleLibrary(lib, version=needVersion)
+    if not found:
+      missing.append((lib, needVersion))
+      continue
+    if needVersion is not None and foundVersion != needVersion:
+      notQA.append((lib, needVersion, foundVersion))
```
Comment (wangcj05) [inline scripts/library_handler.py]: It seems to me, the QA versions are the ones that we included in the config file, and we will not have the ranges for versions of libraries. Let's have an discussion during the stand up meeting today.

File: developer_tools/packaging/pip_ve_osx.sh [hunk 4]
```diff
@@ -86,8 +86,8 @@ virtualenv --always-copy $INSTALL_DIR
 PATH="$OLD_PATH"
 
 source $INSTALL_DIR/bin/activate
-#Call RavenUtils to return the pip install command with the qa'd versions
-`python $SCRIPT_DIR/../../scripts/TestHarness/testers/RavenUtils.py --pip-install`
+#Call library_handler to return the pip install command with the qa'd versions
+`python $SCRIPT_DIR/.bray_handler.py pip --action install`
```
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply developer_tools/packaging/pip_ve_osx.sh]: fixed, thanks

File: framework/Driver.py [hunk 5]
```diff
@@ -105,20 +104,37 @@ def checkVersions():
...
-  import RavenUtils
-  sys.path.pop() #remove testers path
-  missing,outOfRange,notQA = RavenUtils.check_for_missing_modules(False)
-  if len(missing) + len(outOfRange) > 0 and RavenUtils.check_versions():
-    print("ERROR: too old, too new, or missing raven libraries, not running:")
-    for error in missing + outOfRange + notQA:
-      print(error)
+  # library handler is in scripts
+  sys.path.append(os.path.join(frameworkDir, '..', "scripts"))
+  import library_handler as LH
```
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply framework/Driver.py]: moved

File: scripts/TestHarness/testers/RavenErrors.py [hunk 6]
```diff
@@ -82,13 +87,14 @@ def check_runnable(self):
       @ In, None
       @ Out, check_runnable, boolean, If True this test can run.
     """
-    missing, too_old, _ = RavenUtils.check_for_missing_modules()
+    missing, not_qa = library_handler.checkLibraries()
```
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply scripts/TestHarness/testers/RavenErrors.py]: This is a rook requirement right now; it runs pylint on all the checkers it can find and enforces `snake_case`. I don't see a way to disable it just for RAVEN testers at this point.

File: scripts/TestHarness/testers/RavenErrors.py [hunk 7]
```diff
@@ -82,13 +87,14 @@ def check_runnable(self):
...
     if len(missing) > 0:
       self.set_skip('skipped (Missing python modules: '+" ".join(missing)+
                     " PYTHONPATH="+os.environ.get("PYTHONPATH", "")+')')
       return False
-    if len(too_old) > 0:
-      self.set_skip('skipped (Old version python modules: '+" ".join(too_old)+
+    if len(not_qa) > 0:
+      self.set_skip('skipped (incorrect version python modules: ' +
```
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply scripts/TestHarness/testers/RavenErrors.py]: see previous

File: scripts/TestHarness/testers/RavenPython.py [hunk 8]
```diff
@@ -146,21 +152,22 @@ def check_runnable(self):
     if self.specs['requires_swig2'] and not RavenPython.has_swig2:
       self.set_skip('skipped (No swig 2.0 found)')
       return False
-    missing, too_old, _ = RavenUtils.check_for_missing_modules()
+    missing, not_qa = library_handler.checkLibraries()
```
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply scripts/TestHarness/testers/RavenPython.py]: see previous

File: scripts/TestHarness/testers/RavenPython.py [hunk 9]
```diff
@@ -146,21 +152,22 @@ def check_runnable(self):
...
     if len(missing) > 0:
       self.set_fail('skipped (Missing python modules: '+" ".join(missing)+
                     " PYTHONPATH="+os.environ.get("PYTHONPATH", "")+')')
       return False
-    if len(too_old) > 0 and RavenUtils.check_versions():
-      self.set_fail('skipped (Old version python modules: '+" ".join(too_old)+
+    if len(not_qa) > 0 and library_handler.checkVersions():
```
[the remark this answers is not in the document]
Comment (PaulTalbot-INL) [reply scripts/TestHarness/testers/RavenPython.py]: see previous

File: tests/framework/CodeInterfaceTests/Neutrino/Neutrino_Base/sampleNeutrino/1/NeutrinoInput.nescene (hunk 2 above)
Comment (PaulTalbot-INL) [reply tests/framework/CodeInterfaceTests/Neutrino/Neutrino_Base/sampleNeutrino/1/NeutrinoInput.nescene]: fixed

**Comment by alfoa** [comment]
Since we are working on this, we should add documentation in the wiki on how to modify/add new libraries in the config file.

**Comment by PaulTalbot-INL** [comment]
> Since we are working on this, we should add documentation in the wiki on how to modify/add new libraries in the config file.

Yeah, I can do that once it looks like we're ready to merge, or right after it's merged.

**Comment by wangcj05** [comment]
@PaulTalbot-INL  changes are good to me. 
@alfoa two questions for you: 1. with this PR, we will pin our libraries versions, and no ranges of libraries can be accepted.
2. snake case is used in some utils scripts that are used by rook to conduct the regression test. Are you ok with it?

**Comment by PaulTalbot-INL** [comment]
FYI, I am changing the pylinting for the testers in RAVEN from the rook linter to RAVEN linter. This requires rewriting of a lot of variables in the testers.

This isn't very consistent, since they are extensions of rook, not RAVEN, and our linter doesn't cover anything else in `scripts`; however, it will look a little more RAVEN-ish (except the inherited methods from rook).

**Comment by wangcj05** [comment]
@PaulTalbot-INL sorry for the late. Checklist is good to me. In addition, you need to modify the wiki for the updated libraries.

**Comment by PaulTalbot-INL** [comment]
Wiki modified: https://github.com/idaholab/raven/wiki/Developer_Information#adding-new-python-dependencies
